bitfield/impl/src/bitfield_specifier.rs

```diff
@@ -0,0 +1,120 @@
+                    ident: input.ident,
+                    generics: input.generics,
+                    brace_token: data_enum.brace_token,
+                    variants: data_enum.variants,
+                }
+            )
+        }
+        syn::Data::Struct(_) => {
+            bail!(
+                input.ident,
+                "structs are not supported as bitfield specifiers",
+            )
+        }
+        syn::Data::Union(_) => {
+            bail!(
+                input.ident,
+                "unions are not supported as bitfield specifiers",
+            )
+        }
+    }
+}
+
+pub fn generate3(input: syn::ItemEnum) -> syn::Result<TokenStream2> {
+    let enum_ident = &input.ident;
+    let count_variants = input.variants.iter().count();
+    if !count_variants.is_power_of_two() {
+        bail!(
+            input,
+            "BitfieldSpecifier expected a number of variants which is a power of 2",
+        )
+    }
+    // We can take `trailing_zeros` returns type as the required amount of bits.
+    let bits = count_variants.trailing_zeros();
+    let bits_literal = syn::LitInt::new(bits as u64, syn::IntSuffix::None, Span2::call_site());
+
+    // let mut discriminator_mask = Punctuated::<syn::Expr, Token![&&]>::new();
+    let mut check_discriminants_tokens = quote! {};
+    for variant in input.variants.iter() {
+        match &variant.fields {
+            syn::Fields::Named(fields_named) => bail!(fields_named, "named fields in enum variants are not supported"),
+            syn::Fields::Unnamed(fields_unnamed) => bail!(fields_unnamed, "unnamed fields in enum variants are not supported"),
+            syn::Fields::Unit => {
+                let variant_ident = &variant.ident;
+                check_discriminants_tokens.extend(quote_spanned! { variant.span() =>
+                    enum #variant_ident {}
+                    impl checks::CheckDiscriminantInRange for #variant_ident {
+                        type CheckType = [(); ((#enum_ident::#variant_ident as usize) < (0x1 << #bits_literal)) as usize ];
+                    }
+                });
+            },
+        }
+    }
+
+    use crate::ident_ext::IdentExt as _;
+    let impl_mod_name = syn::Ident::from_str(&format!(
+        "impl_bitfield_checks_for_{}",
+        enum_ident.to_owned_string(),
+    ));
+
+    Ok(quote!{
+        mod #impl_mod_name {
+            use super::*;
+
+            #check_discriminants_tokens
+        }
+
+        impl bitfield::Specifier for #enum_ident {
+            const BITS: usize = #bits_literal;
+            type Base = <[(); #bits_literal] as bitfield::SpecifierBase>::Base;
+            type Face = Self;
+        }
+
+        impl FromBits<<#enum_ident as bitfield::Specifier>::Base> for #enum_ident {
+            fn from_bits(base: bitfield::Bits<<#enum_ident as bitfield::Specifier>::Base>) -> Self {
+                unsafe {
+                    std::mem::transmute::<_, _>(base.into_raw())
```

Comment:
As a user I would prefer to see this implemented safely via a `match`.

Reply:
This is one of the things where I was really wondering how the specs (your workshop) wants me to implement this functionality and due to performance reasons I thought that matching the enum could potentially result in sub optimal code in `release` compared to the `transmute` above.

I cannot see a way how the `transmute` above could be abused, yet, I totally agree that I should get rid of it if I can verify with benchmarks that a match results in the same performance.

Reply:
Further note: In the workshop you say that implementers should try to implement the `bitfield` macro in a way that on `--release` it won't differ (much) in the resulting code and should be as fast as handwritten bit manipulation code.
Maybe it would be awesome to have an integrated minor benchmarking suite for the `bitfield` workshop so that users can easily test if they actually reached the goal so that not everybody has to implement their own benchmarking suite and potentially fail at doing so here and there because correct benchmarking is a hard topic on its own (imo).

Reply:
On this stance I am wondering a lot how my abstractions through `PushBits` and `PopBits` are actually good enough to hit the performance in `--release` mode. Probably going to write some benchmarks for it.

Reply:
I have hard coded this in the playground to inspect the generated LLVM assembly and it looks like Rust (or LLVM) is not able to optimize the very simple match into something that is equal to just the transmute in the end.

Look here: https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=2119f66f25b214d1a8bdf26332c3d5b7

The LLVM assembly generated for `day_from_bits` is:

```llvm
; Function Attrs: noinline nonlazybind uwtable
define i32 @day_from_bits(i32) unnamed_addr #4 {
start:
  %switch.tableidx = add i32 %0, -1
  %1 = icmp ult i32 %switch.tableidx, 7
  br i1 %1, label %switch.lookup, label %bb16

bb16:                                             ; preds = %start
; call std::panicking::begin_panic
  tail call fastcc void @_ZN3std9panicking11begin_panic17he8c6c648ff6dd317E()
  unreachable

switch.lookup:                                    ; preds = %start
  ret i32 %0
}
```

And for `day_from_bits_unchecked` it is:

```llvm
; Function Attrs: noinline norecurse nounwind nonlazybind readnone uwtable
define i32 @day_from_bits_unchecked(i32 returned %bits) unnamed_addr #5 {
start:
  ret i32 %bits
}
```

Reply:
I just implemented the match version using `std::hint::unreachable_unchecked` hint.
This actually worked in helping the compiler generate better code so for the Rust playground example we get the following code:

```llvm
; Function Attrs: noinline norecurse nounwind nonlazybind readnone uwtable
define i32 @day_from_bits(i32) unnamed_addr #2 {
start:
  ret i32 %0
}
```